l10n_it_fatturapa_in/tests/test_import_fatturapa_xml.py

```diff
@@ -1006,6 +1006,34 @@ def test_duplicated_vat_on_partners(self):
         attach = self.run_wizard("duplicated_vat", "IT05979361218_012.xml", mode=False)
         self.assertFalse(attach.xml_supplier_id)
         self.assertTrue(attach.inconsistencies)
+        duplicated_supplier.unlink()
+
+    def test_payment_terms_on_partners(self):
+        supplier = self.env["res.partner"].search(
+            [("vat", "=", "IT05979361218")], limit=1
+        )
+        payment_term = self.env["account.payment.term"].create(
+            {
+                "name": "Test",
+                "line_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "value": "balance",
+                            "days": 16,
+                            "option": "after_invoice_month",
+                        },
+                    )
+                ],
+            }
+        )
+        supplier.property_supplier_payment_term_id = payment_term
```

Comment:
Qui viene creato il termine di pagamento.

Cosa succede se non lo crei?   Quando un utente importa la fattura, il sistema non dovrebbe controllare e se mancano i termini di pag. per il supplier dovrebbe fermarsi e dire all'utente di crearli?

Poi, discorso a parte, impostare e far ricalcolare le scadende ad Odoo rischia di ottenere un risultato diverso da quello che (eventualmente)   è indicato in fattura, tipo https://github.com/fredzamoabg/l10n-italy/blob/6942b6d2567e345105c74edc59a251355af91a6a/l10n_it_fatturapa_in/tests/data/IT01234567890_FPR10.xml#L89-L96

Per quanto sia un discorso delicato quello dello scadenziario, si potrebbe valutare se implementare la funzionalità. In pratica si tratta di inserire le righe giuste in fattura in termini di conto di debito vs. fornitori.

Altrimenti, forse sarebbe meglio disabilitare il ricalcolo in presenza delle informazioni di pagamento in fattura, perché l'utente importa la fattura e si trova scadenze diverse da quelli indicate.

Reply:
se il termine di pagamento non è presente sul partner allora `l10n_it_fatturapa_in` lo va a prendere dal XML [qui](https://github.com/OCA/l10n-italy/blob/6a91bfd82a70893d570b5410f9da8c533d279f8b/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py#L1391-L1396)

[qui](https://github.com/OCA/l10n-italy/blob/6a91bfd82a70893d570b5410f9da8c533d279f8b/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py#L1060) invece viene settato in dato in fattura da `l10n_it_fatturapa_in` ed il ricalcolo da `Odoo` chiamerebbe [questo](https://github.com/odoo/odoo/blob/eeca5b27ff2494b4b31038750b487a4b0e5a9d8f/addons/account/models/account_move.py#L953-L961) eseguendo di fatti: https://github.com/odoo/odoo/blob/14.0/addons/account/models/account_move.py#L961

non ho visto problemi a riguardo ma magari mi sono perso qualcosa...

Reply:
Mi ero completamente perso/scodato il pezzo che crea pagamenti.
Ma ricalcolandoli non rischia di cambiare le date?
la `set_payments_data()` viene chiamata prima del ricalcolo forzato.

Per chiarire lo scenario, si tratta per es. di un 30/60/90 dffm.

Reply:
dalle prove che ho fatto non ho visto problemi con le date, ai qualche suggerimento da poter includere nei test?

Reply:
Potresti mettere un test per es. con data fattura 28/3, pagamenti 30/4, 30/5, 30/6.  Notare il 30/5.

Poco importa che il cliente abbia generato una scadenza non corretta nel suo XML, per come la vedo io Odoo la deve preservare e non sistamarla spostandola al 31/5.

Il mio sospetto è che la recompute sposti la data del pagamento.

Inoltre mi sa che la ricerca dei termini di pagamento non avviene per date, solo per tipologia (per es. "bonifico"). Se il cliente inserisce date a caso (cosa inusuale ma che nessuno vieta) per i pagamenti, come fa il sistema a trovare una voce nelle condizioni di pagamento che probabilmente non esiste? con quelle scadenze specifiche intendo.

Reply:
> Poco importa che il cliente abbia generato una scadenza non corretta nel suo XML, per come la vedo io Odoo la deve preservare e non sistamarla spostandola al 31/5.

questo non lo capisco, da [qui](https://github.com/OCA/l10n-italy/blob/6a91bfd82a70893d570b5410f9da8c533d279f8b/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py#L1391-L1396) è chiaro che il settaggio del campo `invoice_date_due` dal file XML avviene solo in una condizione ed Odoo lo cambia solo nella condizione opposta ([qui](https://github.com/odoo/odoo/blob/eeca5b27ff2494b4b31038750b487a4b0e5a9d8f/addons/account/models/account_move.py#L958)); questo vuol dire che se il campo viene settato da `l10n_it_fatturapa_in` Odoo non lo cambierà.

per il resto vedrò asap di inserire il test, grazie

Reply:
ma io non mi riferisco a `invoice_date_due`, mi riferisco a https://github.com/odoo/odoo/blob/00360b4dd321e1164ec39c4a2c8a2811a0ac1222/addons/account/models/account_move.py#L3570

30/5 è la scadenza del secondo pagamento, `invoice_date_due` sarebbe 30/6 in questo caso.

Ho in mente scadenze multiple dei pagamenti, tipo 30/60/90, e non una singola scadenza di tutta la fattura. Il cliente - sotto effetto di qualche acido - potrebbe aver impostato come scadenze dei pagamenti 27 marzo, 19 aprile, 22 maggio. Se si forza un ricalcolo sempre e comunque, io non vedo nulla (almeno a prima vista) che impedisca appunto il ricalcolo dei pagamenti, con date diverse da quelle dell'XML originario.

Reply:
@TheMule71 in caso affermativo, per favore, modifica la tua review in **Request changes**.

Reply:
Da verificare se succede. Bisognerebbe fare un test.

E comunque non c'è un consenso comune sul comportamento da tenere per fatture passive "strane".

Un partito è per importarle così come sono rimanendo il più aderenti possibili a quanto transitato per lo SdI.
L'altro partito è per imporre le condizioni contrattuali, ignorando quanto scritto in fattura dal cliente se in contrasto con le stesse.

Qui è analogo, se 30/60/90 sono le scadenze concordate, e Odoo le calcola secondo quanto concordato, le date presenti nell'XML "vincono" o no a seconda del partito in cui militi.

Io non sto da nessuna parte, mi limito solo a segnalare se c'è un punto che potrebbe essere critico/